File: internal/eventlog/position.go
package eventlog

import (
	"cmp"
	"fmt"
	"slices"
	"strconv"
	"strings"
)

// Position is the global ordering key. Every replica sorts the union of every
// log file by (ts, rep, seq) and gets the same sequence; that is the whole
// merge algorithm.
type Position struct {
	TS  int64
	Rep string
	Seq uint64
}

// Position returns e's ordering key.
func (e Envelope) Position() Position {
	return Position{TS: e.TS, Rep: e.Rep, Seq: e.Seq}
}

// Compare orders p against q, returning -1, 0 or 1.
func (p Position) Compare(q Position) int {
	if c := cmp.Compare(p.TS, q.TS); c != 0 {
		return c
	}
	if c := strings.Compare(p.Rep, q.Rep); c != 0 {
		return c
	}
	return cmp.Compare(p.Seq, q.Seq)
}

// Less reports whether p sorts before q.
func (p Position) Less(q Position) bool { return p.Compare(q) < 0 }

// String encodes p as "<ts>-<rep>-<seq>", URL-safe because rep is base62 and
// the other two are decimal.
//
// The encoding is a cursor, not a sort key: compare positions by parsing them,
// not by comparing the strings.
func (p Position) String() string {
	return fmt.Sprintf("%d-%s-%d", p.TS, p.Rep, p.Seq)
}

// ParsePosition decodes the encoding produced by Position.String.
func ParsePosition(s string) (Position, error) {
	parts := strings.Split(s, "-")
	if len(parts) != 3 {
		return Position{}, fmt.Errorf("eventlog: %q is not a position (want <ts>-<rep>-<seq>)", s)
	}
	ts, err := strconv.ParseInt(parts[0], 10, 64)
	if err != nil || ts <= 0 {
		return Position{}, fmt.Errorf("eventlog: %q has no valid ts", s)
	}
	if !ValidReplicaID(parts[1]) {
		return Position{}, fmt.Errorf("eventlog: %q has no valid rep", s)
	}
	seq, err := strconv.ParseUint(parts[2], 10, 64)
	if err != nil || seq == 0 {
		return Position{}, fmt.Errorf("eventlog: %q has no valid seq", s)
	}
	return Position{TS: ts, Rep: parts[1], Seq: seq}, nil
}

// Sort orders evs by position. The order is total — no two events share a
// position unless a replica has repeated a seq — so it is stable under shuffle.
func Sort(evs []Envelope) {
	slices.SortStableFunc(evs, func(a, b Envelope) int {
		return a.Position().Compare(b.Position())
	})
}

File: internal/web/assets/js/position.mjs
/*
  Log positions — the dashboard's event cursor.

  Every event the server sends carries `position`, the string form of the
  log position (ts, rep, seq) that internal/eventlog/position.go defines.
  That triple is what every replica agrees on, so it is what `?at=`,
  `?since=` and the replay buffer address events by. The event's `id` is the
  server cache's row id: a DOM key, renumbered by any rebuild, and never a
  cursor.

  The encoding is a cursor, not a sort key. "1000-aaaaaa-9" and
  "1000-aaaaaa-10" compare the wrong way as strings — always parse first and
  use comparePositions.

  A position with an empty replica ("<ts>--<n>") is a legacy cursor: a row
  translated from a pre-log database, addressed by its cache row id because
  it has no replica or seq of its own. It is meaningful only within one
  cache. The empty replica sorts before every real one, which matches the
  server.
*/

// parsePosition decodes "<ts>-<rep>-<seq>" into a comparable triple, or
// returns null when the input is not a position.
export function parsePosition(s) {
  if (typeof s !== "string" || s === "") return null;
  const parts = s.split("-");
  if (parts.length !== 3) return null;
  if (!/^\d+$/.test(parts[0]) || !/^\d+$/.test(parts[2])) return null;
  if (parts[1] !== "" && !/^[0-9A-Za-z]+$/.test(parts[1])) return null;
  const ts = Number(parts[0]);
  const seq = Number(parts[2]);
  if (!(ts > 0) || !(seq > 0)) return null;
  return { ts, rep: parts[1], seq };
}

// isPosition reports whether s parses as a position.
export function isPosition(s) {
  return parsePosition(s) !== null;
}

// comparePositions orders two encoded positions, returning -1, 0 or 1.
// A null / absent / unparseable cursor sorts before every real position —
// that is the "genesis" end of the log, which is what an empty ?at means.
export function comparePositions(a, b) {
  const pa = parsePosition(a);
  const pb = parsePosition(b);
  if (pa === null && pb === null) return 0;
  if (pa === null) return -1;
  if (pb === null) return 1;
  if (pa.ts !== pb.ts) return pa.ts < pb.ts ? -1 : 1;
  if (pa.rep !== pb.rep) return pa.rep < pb.rep ? -1 : 1;
  if (pa.seq !== pb.seq) return pa.seq < pb.seq ? -1 : 1;
  return 0;
}

// samePosition is equality on the parsed triple, so an unparseable value
// never accidentally equals a real cursor.
export function samePosition(a, b) {
  if (!isPosition(a) || !isPosition(b)) return false;
  return comparePositions(a, b) === 0;
}

// sortByPosition orders a list of events ascending by their cursor. The
// server already returns them in this order; the buffer re-sorts because a
// live frame can arrive out of band.
export function sortByPosition(events) {
  return events
    .slice()
    .sort((a, b) => comparePositions(a.position, b.position));
}
